Added ZeroTier 1.2.12 package

joseph-henry opened the pull request:
_Motivation:_ ZeroTier delivers the capabilities of VPNs, SDN, and SD-WAN with a single system. Manage all your connected resources across both local and wide area networks as if the whole world is a single data center.

### Checklist
- [x] Build rule `all-supported` completed successfully
- [x] Package upgrade completed successfully
- [x] New installation of package completed successfully


**Build rule `all-supported` completed successfully**
 - The following packages completed successfully: `alpine`, `armada370`, `armada375`, `armada38x`, `armadaxp`, `comcerto2k`, `denverton`, `dockerx64`, `evansport`, `kvmx64`, `monaco`, `rtd1296`, `x64`.
 - The following packages did **not** complete successfully due to toolchains that do not support C++ features required for ZeroTier: `hi3535`

**Package upgrade completed successfully**
ZeroTier possesses its own (optional) in-band update mechanism which operates over its own encrypted connection. It is currently disabled for Synology NAS builds but will be enabled when enough testing has been concluded. Alternatively the aforementioned update mechanism may be supplanted by the native system update mechanism, but this has not been fully decided on yet. Feedback is welcome.

**New installation of package completed successfully**
Tested on `DS216+II`. Note: The ZeroTier service has been tested and works on most Synology models. Only this package is new.

***

Other notes: The GUI is available but no shortcut appears in the Main Menu of installed applications. Until this is addressed all operations can be completed via the CLI interface using `zerotier-cli`.

ymartin59 commented on cross/zerotier/Makefile:
```diff
@@ -0,0 +1,30 @@
+
+GNU_CONFIGURE = 1
+CONFIGURE_ARGS = HAVE_CXX=yes
+CONFIGURE_TARGET = myConfigure
```
To ease debug you should name Makefile targets "zerotier_custom_configure" and "zerotier_custom_install"

ymartin59 commented on spk/zerotier/Makefile:
```diff
@@ -0,0 +1,35 @@
+STARTABLE = yes
+
+LICENSE  = GPLv3
+INSTALLER_SCRIPT = src/installer.sh
```
To reduce script volume, you should take benefits from generic installer and startup script: https://github.com/SynoCommunity/spksrc/wiki/Service-Support
For security reason, service should not run as root if not required.

ymartin59 commented on spk/zerotier/src/dsm-control.sh:
```diff
@@ -0,0 +1,85 @@
+DNAME="ZeroTier"
+PACKAGE="zerotier"
+INSTALL_DIR="/usr/local/${PACKAGE}"
+ZT_HOME_DIR="/var/lib/zerotier-one"
```
This should be `${SYNOPKG_PKGDEST}/var` aka `/var/packages/zerotier/target/var`

ymartin59 commented on spk/zerotier/src/dsm-control.sh:
```diff
@@ -0,0 +1,85 @@
+ZEROTIER="${INSTALL_DIR}/bin/zerotier-one"
+WATCHDOG="${INSTALL_DIR}/bin/zerotier-watchdog.sh"
+PID_FILE="${ZT_HOME_DIR}/zerotier-one.pid"
+LOG_FILE="/var/log/zerotier-one.log"
```
The same way, logs should go into `/var/packages/zerotier/target/var`. By the way, generic service handles logging for you.

ymartin59 commented on spk/zerotier/src/dsm-control.sh:
```diff
@@ -0,0 +1,85 @@
+
+start_daemon ()
+{
+    ${WATCHDOG} stop ;
```
I do not understand interest of watchdog process in addition to service itself. What does it run for?

joseph-henry commented:
Thanks, I'm working on addressing these issues now. But I've got two questions:

 - We use the custom `/var/lib/zerotier-one` path to store cryptographic identity files. Now, for convenience I typically allow our `identity.*` files to persist even after uninstall. It can make re-joining your network much easier in the future if the application is installed again and these files are negligible in size and do not grow. Is there another place I should put these instead?

- Even after setting `SERVICE_PORT` and `PKG_ICON` I still don't see a shortcut in the Main Menu, is there something I'm missing?

joseph-henry commented:
At this stage I'm just looking for a persistent place to keep identity files (very tiny!) between reinstalls. I *think* I've addressed the requested changes. Let me know if there's anything else I should take care of.

Thanks!

joseph-henry commented on spk/zerotier/Makefile:
```diff
@@ -0,0 +1,39 @@
+SPK_NAME = zerotier
+SPK_VERS = 1.2.12
+SPK_REV = 0
```
Changed.

joseph-henry commented on spk/zerotier/Makefile:
```diff
@@ -0,0 +1,39 @@
+DSM_UI_DIR = ui
+RELOAD_UI = yes
+OS_MIN_VERS = 6.0-7321
+FIRMWARE = 6.0-7321
```
Removed. I had these in here because during app submission with Synology they required them to be set.

joseph-henry commented on cross/zerotier/Makefile:
```diff
@@ -0,0 +1,36 @@
+
+GNU_CONFIGURE = 1
+CONFIGURE_ARGS = HAVE_CXX=yes
+CONFIGURE_TARGET = zerotier_custom_configure
```
I've removed the empty target, but when I don't specify a `CONFIGURE_TARGET` I get the following error:

```
env: ./configure: No such file or directory
../../mk/spksrc.configure.mk:50: recipe for target 'configure_target' failed
make[1]: *** [configure_target] Error 127
make[1]: Leaving directory '/spksrc/cross/zerotier'
../../mk/spksrc.depend.mk:44: recipe for target 'depend_target' failed
make: *** [depend_target] Error 2
```

joseph-henry commented on cross/zerotier/Makefile:
```diff
@@ -0,0 +1,36 @@
+INSTALL_TARGET = zerotier_custom_install
+
+ENV += ZT_SYNOLOGY=1
+REQUIRED_DSM = 6.1
```
Moved. Thank you.

joseph-henry commented:
Thanks, I will ignore the warnings about libstdc++ since ZeroTier seems to work very well on Synology NASs and has for a number of years.

However, I am somewhat confused about the phrase: 

> A packager may consider running such a user interface as root is a non-affordable security risk and will push effort to get it run as non-privileged user.

ZeroTier must run as root due to its use of the kernel TUN module, does this mean I cannot use the generic service/installer script?

Also, I have had ZeroTier running for a few years using the current service/install script, I am unsure what benefit there will be in using the generic one.

I appreciate all of your help.
